app/page.tsx

```diff
@@ -1,98 +1,77 @@
-// Homepage - User login and signup
 "use client";
-import React, { ChangeEventHandler, FormEventHandler, useState } from "react";
-import { signIn } from "next-auth/react";
+
+import { useEffect, useState, useRef } from "react";
+import { AuthorizeOptions } from "@/lib/zoomapi";
+import { signIn, useSession } from "next-auth/react";
+import { SessionProvider } from "next-auth/react";
+import { ZoomApiWrapper } from "@/lib/zoomapi";
 import { useRouter } from "next/navigation";
+import { log, Action } from "@/lib/log";
 import Alert from "@/components/Alert";
-import { Action, log } from "@/lib/log";
+import { LoadingComponent } from "@/components/LoadingComponent";
 
-const Login = () => {
+function App() {
   const [error, setError] = useState("");
-  const [userInfo, setUserInfo] = useState({
-    email: "",
-    password: "",
-  });
+  const { data: session } = useSession();
+  const effectRan = useRef(false);
+
   const router = useRouter();
-  const { email, password } = userInfo;
 
-  const handleChange: ChangeEventHandler<HTMLInputElement> = ({ target }) => {
-    const { name, value } = target;
-    setUserInfo({ ...userInfo, [name]: value });
-  };
+  async function getZoomApi(): Promise<ZoomApiWrapper> {
+    const zoomModule = await import("@/lib/zoomapi");
+    return zoomModule.zoomApi;
+  }
 
-  const handleSubmit: FormEventHandler<HTMLFormElement> = async (e) => {
-    e.preventDefault();
+  useEffect(() => {
+    if (effectRan.current) return;
 
-    const res = await signIn("credentials", {
-      email,
-      password,
-      redirect: false,
-    });
+    effectRan.current = true;
 
-    if (res?.error) {
-      log(Action.LOG_IN_FAIL, email, res);
-      return setError(res.error);
+    async function handleAuth() {
+      try {
+        console.log("handleAuth called");
+        const zoomApi = await getZoomApi();
+        const options: AuthorizeOptions = {
+          state: "state",
+          codeChallenge: "codeChallenge",
+        };
+        const res = await zoomApi.authorize(options);
+        console.log("zoom sdk authorized:", res);
+        zoomApi.setAuthorizeCallback(async (event) => {
+          console.log("zoom api authorize callback called:", event);
+          const res = await signIn("credentials", {
+            code: event.code,
+            redirect: false,
+          });
+          if (res?.error) {
+            console.error("next auth sign in failed:", res.error);
+            return setError(res.error);
+          }
+          router.replace("/main");
+          if (session?.user?.email) {
+            log(Action.LOG_IN, session.user.email);
+          }
+        });
+      } catch (error) {
+        console.error("Failed to authorize zoom API:", error);
+        setError((error as Error).message);
```

Comment:
Note that in TS, you can put a type annotation on the caught exception to avoid needing to cast it, like:

```suggestion
      } catch (error: Error) {
        console.error("Failed to authorize zoom API:", error);
        setError(error.message);
```

Reply:
In `catch` statement, the variable type annotation MUST be `any`.

```
Type error: Catch clause variable type annotation must be 'any' or 'unknown' if specified.
```

Reply:
Ah, OK, good to know.